File: src/backend/consumer_service/services/order.py
from fastapi import HTTPException
import requests
import os

class OrderService:
    def __init__(
        self,
        user=None,
        id=None,
        pyxiId=None,
        senderId=None,
        receiverId=None,
        description=None,
        itemId=None,
        reason=None,
        canceledBy=None,
    ):
        self.id = id
        self.user = user
        self.url = f'{os.getenv("AWS_HOST", "http://localhost")}:3000'
        self.prefix = "/orders"

        self.pyxiId = pyxiId
        self.senderId = senderId
        self.receiverId = receiverId
        self.description = description
        self.itemId = itemId

        self.reason = reason
        self.canceledBy = canceledBy

    async def get_receiver_orders(self) -> list:
        response = requests.get(
            self.url + self.prefix + "/receiver/" + self.user,
            headers={"Accept": "application/json", "Content-Type": "application/json"},
        )

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(
                status_code=response.status_code, detail="Errei fui mlk"
            )

    async def get_sender_orders(self):
        response = requests.get(
            self.url + self.prefix + "/sender/" + self.user,
            headers={"Accept": "application/json", "Content-Type": "application/json"},
        )

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(
                status_code=response.status_code, detail="Errei fui mlk"
            )

    async def get_all_pendings(self):
        response = requests.get(
            self.url + self.prefix + "/pending",
            headers={"Accept": "application/json", "Content-Type": "application/json"},
        )

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(
                status_code=response.status_code, detail="Errei fui mlk"
            )

    async def get_order_details(self):
        response = requests.get(
            self.url + self.prefix + "/" + self.id,
            headers={"Accept": "application/json", "Content-Type": "application/json"},
        )

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(
                status_code=response.status_code, detail="Errei fui mlk"
            )

    async def get_last_pending(self):
        response = requests.get(
            self.url + self.prefix + "/last_pending",
            headers={"Accept": "application/json", "Content-Type": "application/json"},
        )

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(
                status_code=response.status_code, detail="Errei fui mlk"
            )

    def create(self, problem: str):
        response = requests.post(
            self.url + self.prefix + f"/{problem}",
            headers={"Accept": "application/json", "Content-Type": "application/json"},
            json={
                "pyxiId": self.pyxiId,
                "senderId": self.senderId,
                "description": self.description,
                "itemId": self.itemId,
            },
        )
        print(response.status_code)
        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(status_code=response.status_code, detail="Errei")

    def accept(self):
        response = requests.put(
            self.url + self.prefix + f"/accept/{self.id}?receiverId={self.receiverId}",
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
        )
        print(response.status_code)

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(status_code=response.status_code, detail="Errei")

    def done(self):
        response = requests.put(
            self.url + self.prefix + f"/close/{self.id}",
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
        )
        print(response.status_code)

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(status_code=response.status_code, detail="Errei")

    def cancel(self):
        response = requests.put(
            self.url + self.prefix + f"/cancel/{self.id}",
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
            json={
                "reason": self.reason,
                "userId": self.canceledBy,
            },
        )

        print(response.status_code)

        if response.status_code == 200:
            data = response.json()

            return data
        else:
            raise HTTPException(status_code=response.status_code, detail="Errei")

File: src/frontend/mobile/lib/api/order.dart
import 'dart:convert';
import 'package:PharmaControl/models/order.dart';
import 'package:http/http.dart' as http;
import 'package:PharmaControl/globals.dart' as globals;

import '../models/pyxis.dart';

var baseurl = globals.baseurl;
var pubUrl = globals.publisherUrl;

Future<List<Order>?> getOrders() async {
  String userUrl = globals.user!.role == "NURSE"
      ? "sender/${globals.user!.id}"
      : "receiver/${globals.user!.id}";

  var response = await http.get(
    Uri.parse("$baseurl/orders/$userUrl"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
  );

  if (response.statusCode == 200) {
    final List<dynamic> data = json.decode(utf8.decode(response.bodyBytes));

    return data.map((d) => Order.fromJson(d)).toList();
  } else {
    return null;
  }
}

Future<List<Pyxis>?> getPyxis() async {
  var response = await http.get(
    Uri.parse("$baseurl/pyxis/"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
  );

  if (response.statusCode == 200) {
    final List<dynamic> data =
        json.decode(utf8.decode(response.bodyBytes))["pyxis"];

    return data.map((d) => Pyxis.fromJsonWithItems(d)).toList();
  } else {
    return null;
  }
}

Future<Order?> fetchLastOrder() async {
  var response = await http.get(
    Uri.parse("$baseurl/orders/last_pending"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
  );

  if (response.statusCode == 200) {
    final data = json.decode(utf8.decode(response.bodyBytes));

    return Order.fromJson(data);
  } else {
    return null;
  }
}

Future<bool> createOrder(String userId, String pyxiId, String problem,
    String? description, String? itemId) async {
  final Map<String, String?> data = {
    "sender_userId": userId,
    "pyxiId": pyxiId,
    "problem": problem,
    "description": description,
    "itemId": itemId,
  };

  var response = await http.post(
    Uri.parse("$pubUrl/order/"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
    body: jsonEncode(data),
  );

  return (response.statusCode == 200);
}

Future<bool> acceptOrder(String userId, String orderId) async {
  final Map<String, String> data = {
    "receiver_userId": userId,
    "order_id": orderId,
  };

  var response = await http.post(
    Uri.parse("$pubUrl/order/accept"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
    body: jsonEncode(data),
  );

  return (response.statusCode == 200);
}

Future<bool> doneOrder(String orderId) async {
  final Map<String, String> data = {
    "order_id": orderId,
  };

  var response = await http.post(
    Uri.parse("$pubUrl/order/done"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
    body: jsonEncode(data),
  );

  return (response.statusCode == 200);
}

Future<bool> cancelOrder(String orderId, String reason, String userId) async {
  final Map<String, String> data = {
    "order_id": orderId,
    "canceled_reason": reason,
    "canceled_userId": userId,
  };

  var response = await http.post(
    Uri.parse("$pubUrl/order/cancel"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
    body: jsonEncode(data),
  );

  return (response.statusCode == 200);
}

Future<bool> rejectOrder(String orderId) async {
  final Map<String, String> data = {
    "order_id": orderId,
  };

  var response = await http.post(
    Uri.parse("$pubUrl/order/reject"),
    headers: <String, String>{'Content-Type': 'application/json;charset=utf-8'},
    body: jsonEncode(data),
  );

  return (response.statusCode == 200);
}
